clients/drcachesim/tracer/raw2trace.h

```diff
@@ -1066,12 +1066,26 @@ template <typename T> class trace_converter_t {
                             int_modoffs, cur_modoffs);
                 // Because we increment the instr fetch first, the signal modoffs may be
                 // less than the current for a memref fault.
-                if (int_modoffs == cur_modoffs ||
+                // It might also be 0, which means it should be on the first instruction.
+                if (int_modoffs == 0 || int_modoffs == cur_modoffs ||
                     int_modoffs + instr_length == cur_modoffs) {
                     impl()->log(4, "Signal/exception interrupted the bb @ +" PIFX "\n",
                                 int_modoffs);
                     append = true;
                     *interrupted = true;
+                    if (int_modoffs == 0) {
+                        // This happens on rseq aborts, where the trace instru includes
+                        // the rseq committing store before the native rseq execution
+                        // hits the abort.  Pretend the abort happened *before* the
```

Comment:
Here is what happens:

1) Instrumented rseq code runs and makes it past the committing store (nop-ed out in reality, but put into the trace).
2) Now DR does the native resq execution.  If that aborts, the abort handler will run, which will probably restart the whole sequence, going back to 1) and recording the rseq sequence again into the trace buffer.  To emulate what would happen natively, we do *not* want the first committing store from the aborted run in the trace: if it had really executed the store it would not have aborted.  Now, the trace instruction entries are for *fetches* and not *retirements*, so maybe you could argue it's ok if the committing store is not a memory instruction: but it almost always is and we don't want its memrefs in the trace.

Does that make it clearer?

Reply:
> iirc we aren't emulating the branches, reads, and stores for the native re-run, are we?

Yes we are.  On the native abort, DR starts "interpreting" anew at the abort handler.  If it restarts (like most do), the re-run will all go into the trace.

Reply:
Do you have a suggestion for tweaking the comment to make it clearer?

Reply:
I must have forgotten that an abort+restart of the native *rseq* would result in going back to the instrumented code, I thought it stayed native until the sequence completed without aborting. Further, since I didn't consider a native rseq abort mattering, I thought the comment "This happens on rseq aborts," was somehow talking about an abort during the instrumented version of the sequence.

Can you add the word "native" to the relevant parts of this comment?

Reply:
Done.